File: app/src/main/java/com/jpcn/chatapp/ui/activities/MessageActivity.java
package com.jpcn.chatapp.ui.activities;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.jpcn.chatapp.R;
import com.jpcn.chatapp.model.Chat;
import com.jpcn.chatapp.model.User;
import com.jpcn.chatapp.ui.adapters.MessageAdapter;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;

public class MessageActivity extends AppCompatActivity {

    private CircleImageView profileImage;
    private TextView username;
    private String userId;

    private FirebaseUser firebaseUser;
    private DatabaseReference reference;

    private ImageButton buttonSend;
    private EditText textSend;

    private MessageAdapter messageAdapter;
    private List<Chat> chats;

    private RecyclerView recyclerView;

    private Intent intent;

    boolean notify = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_message);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle("");
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(MessageActivity.this, MainActivity.class).setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP));
            }
        });

        profileImage = findViewById(R.id.profileImage);
        username = findViewById(R.id.username);
        buttonSend = findViewById(R.id.buttonSend);
        textSend = findViewById(R.id.textSend);

        recyclerView = findViewById(R.id.recyclerViewMessages);
        recyclerView.setHasFixedSize(true);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getApplicationContext());
        linearLayoutManager.setStackFromEnd(true);
        recyclerView.setLayoutManager(linearLayoutManager);

        intent = getIntent();
        userId = intent.getStringExtra("userId");

        buttonSend.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                notify = true;
                String message = textSend.getText().toString();
                if (!message.isEmpty()) {
                    sendMessage(firebaseUser.getUid(), userId, message);
                } else {
                    Toast.makeText(MessageActivity.this, "You can't send empty message", Toast.LENGTH_SHORT).show();
                }
                textSend.setText("");
            }
        });

        firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
        reference = FirebaseDatabase.getInstance().getReference("users").child(userId);

        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                User user = dataSnapshot.getValue(User.class);
                assert user != null;
                username.setText(user.getUsername());
                if (user.getImageURL().equals("default")) {
                    profileImage.setImageResource(R.mipmap.ic_launcher);
                } else {
                    Picasso.get().load(user.getImageURL()).into(profileImage);
                }

                readMessage(firebaseUser.getUid(), userId, user.getImageURL());
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    private void sendMessage(String sender, String receiver, String message) {
        DatabaseReference reference = FirebaseDatabase.getInstance().getReference();

        HashMap<String, Object> hashMap = new HashMap<>();
        hashMap.put("sender", sender);
        hashMap.put("receiver", receiver);
        hashMap.put("message", message);

        reference.child("Chats").push().setValue(hashMap);

        putChatUser(receiver, sender, message);
        putChatUser(sender, receiver, message);
    }

    private void putChatUser(final String receiver, final String sender, final String message) {
        final DatabaseReference chatsReference = FirebaseDatabase.getInstance().getReference("ChatUser")
                .child(receiver)
                .child(sender);

        chatsReference.addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (!dataSnapshot.exists()) {
                    chatsReference.child("id").setValue(sender);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

        reference = FirebaseDatabase.getInstance().getReference("users").child(firebaseUser.getUid());
        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    private void readMessage(final String myid, final String userid, final String imageurl) {
        chats = new ArrayList<>();
        reference = FirebaseDatabase.getInstance().getReference("Chats");
        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                chats.clear();
                for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                    Chat chat = snapshot.getValue(Chat.class);
                    if (chat.getReceiver().equals(myid) && chat.getSender().equals(userid) ||
                            chat.getReceiver().equals(userid) && chat.getSender().equals(myid)) {
                        chats.add(chat);
                    }
                }

                messageAdapter = new MessageAdapter(MessageActivity.this, chats, imageurl);
                recyclerView.setAdapter(messageAdapter);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

    }

    private void status(String status) {
        reference = FirebaseDatabase.getInstance().getReference("users").child(firebaseUser.getUid());

        HashMap<String, Object> hashMap = new HashMap<>();
        hashMap.put("status", status);

        reference.updateChildren(hashMap);
    }

    @Override
    protected void onResume() {
        super.onResume();
        status("online");
    }

    @Override
    protected void onPause() {
        super.onPause();
        status("offline");
    }
}

File: app/src/main/java/com/jpcn/chatapp/ui/activities/MessageActivity.kt
package com.jpcn.chatapp.ui.activities

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.jpcn.chatapp.R
import com.jpcn.chatapp.databinding.ActivityMessageBinding
import com.jpcn.chatapp.model.Chat
import com.jpcn.chatapp.model.User
import com.jpcn.chatapp.ui.adapters.MessageAdapter
import com.squareup.picasso.Picasso

class MessageActivity : AppCompatActivity() {
    var notify = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivityMessageBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupToolbar(binding)
        setupRecyclerVIew(binding)

        val userId = intent.getStringExtra("userId").orEmpty()

        FirebaseAuth.getInstance().currentUser?.let { firebaseUser ->
            setupSendMessage(binding, firebaseUser, userId)
            setupUserInfo(userId, binding, firebaseUser)
        }
    }

    private fun setupSendMessage(
        binding: ActivityMessageBinding,
        firebaseUser: FirebaseUser,
        userId: String
    ) {
        binding.buttonSend.setOnClickListener {
            notify = true
            val message = binding.textSend.text.toString()
            if (message.isNotBlank()) {
                sendMessage(firebaseUser.uid, userId, message)
            } else {
                Toast.makeText(this@MessageActivity, "You can't send empty message", Toast.LENGTH_SHORT).show()
            }
            binding.textSend.setText("")
        }
    }

    private fun setupUserInfo(
        userId: String,
        binding: ActivityMessageBinding,
        firebaseUser: FirebaseUser
    ) = FirebaseDatabase.getInstance().getReference("users").child(userId)
        .addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                dataSnapshot.getValue(User::class.java)?.let { user ->
                    binding.username.text = user.username
                    if (user.imageURL == "default") {
                        binding.profileImage.setImageResource(R.mipmap.ic_launcher)
                    } else {
                        Picasso.get().load(user.imageURL).into(binding.profileImage)
                    }
                    readMessage(firebaseUser.uid, userId, user.imageURL, binding)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) = Unit
        })

    private fun setupRecyclerVIew(binding: ActivityMessageBinding) {
        binding.recyclerViewMessages.setHasFixedSize(true)
        val linearLayoutManager = LinearLayoutManager(applicationContext)
        linearLayoutManager.stackFromEnd = true
        binding.recyclerViewMessages.layoutManager = linearLayoutManager
    }

    private fun setupToolbar(binding: ActivityMessageBinding) {
        setSupportActionBar(binding.toolbar)
        supportActionBar?.title = ""
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        binding.toolbar.setNavigationOnClickListener {
            startActivity(
                Intent(
                    this@MessageActivity,
                    MainActivity::class.java
                ).setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
            )
        }
    }

    private fun sendMessage(sender: String, receiver: String?, message: String) {
        val reference = FirebaseDatabase.getInstance().reference
        val hashMap = HashMap<String, Any?>()
        hashMap["sender"] = sender
        hashMap["receiver"] = receiver
        hashMap["message"] = message
        reference.child("Chats").push().setValue(hashMap)
        putChatUser(receiver, sender, message)
        putChatUser(sender, receiver, message)
    }

    private fun putChatUser(receiver: String?, sender: String?, message: String) {
        val chatsReference = FirebaseDatabase.getInstance().getReference("ChatUser")
            .child(receiver!!)
            .child(sender!!)
        chatsReference.addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (!dataSnapshot.exists()) {
                    chatsReference.child("id").setValue(sender)
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
        FirebaseAuth.getInstance().currentUser?.let { user ->
            val reference = FirebaseDatabase.getInstance().getReference("users").child(user.uid)
            reference.addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {}
                override fun onCancelled(databaseError: DatabaseError) {}
            })
        }
    }

    private fun readMessage(userId: String, userid: String?, imageUrl: String?, binding: ActivityMessageBinding) {
        val chats = arrayListOf<Chat>()
        val reference = FirebaseDatabase.getInstance().getReference("Chats")
        reference.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                chats.clear()
                for (snapshot in dataSnapshot.children) {
                    val chat = snapshot.getValue(Chat::class.java)
                    if (chat?.receiver == userId && chat.sender == userid ||
                        chat?.receiver == userid && chat?.sender == userId
                    ) {
                        chats.add(chat)
                    }
                }
                val messageAdapter = MessageAdapter(this@MessageActivity, chats, imageUrl)
                binding.recyclerViewMessages.adapter = messageAdapter
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    private fun status(status: String) {
        FirebaseAuth.getInstance().currentUser?.let { user ->
            val reference = FirebaseDatabase.getInstance().getReference("users").child(user.uid)
            val hashMap = HashMap<String, Any>()
            hashMap["status"] = status
            reference.updateChildren(hashMap)
        }
    }

    override fun onResume() {
        super.onResume()
        status("online")
    }

    override fun onPause() {
        super.onPause()
        status("offline")
    }
}